app/processing/index.js

```diff
@@ -1,20 +1,22 @@
 const { processingConfig } = require('../config')
-
+const waitForIdleMessaging = require('../messaging/wait-for-idle-messaging')
 const schedulePendingSettlements = require('./schedule')
-const { getStatement, sendStatement } = require('./statement')
+const { getStatement, sendStatement, validateStatement } = require('./statement')
+const updateScheduleByScheduleId = require('./statement/update-schedule-by-schedule-id')
 
 const start = async () => {
   try {
     if (processingConfig.constructionActive) {
       const pendingStatements = await schedulePendingSettlements()
-      if (!pendingStatements) {
-        throw new Error('No statements to be generated')
-      }
 
       for (const pendingStatement of pendingStatements) {
         try {
+          await waitForIdleMessaging()
```

Comment:
Adding wait here to reduce risk of invalid statements.

Now we're dealing with PPA, there may be multiple settlements settled at the same time that need to be included in the statement construction to ensure the values are correct.  However, there is no way for the service to know ahead of time what settlements it needs.

As settlements arrive as separate messages there is a risk that construction begins for settlement A before settlement B has been saved to the database.  Result = incorrect statement.

So as a safety check, before generating a statement, ensure that all message subscriptions are empty with no messages yet to be delivered to the service.  This significantly reduces the risk.  Especially when coupled with a "cool off period" of statement construction based on the `received` date.  ie wait `5` minutes after a settlement is received before processing it.

Originally had this call outside the loop, but found that when there are a lot of pending schedules that take a while to construct, messages can start to stream into the service and saw a manifestation of the risk I was trying to avoid.  Therefore the check is now performed before every statement construction.

Not an issue now, but we should at some point look to refine this service to support  autoscaling so that safeguards like this don't slow down performance.  Not going to be an issue for current volumes though.

Reply:
In a situation where we have large volume incoming message, I am a bit concerned about performance issue and how long we can keep active, a single sql transaction connection.

Reply:
This isn't within a transaction so that isn't an issue, however, you have a point as this does risk the 5 minute grace period being hit.  Not a problem with a single replica, but would be if more.

I think it might be better to just take this out and just up the settlement wait time.

Reply:
I've opted to move it outside the loop to avoid the above issue.  That way the service if effectively idle if there are active messages.

Reason for keeping is I worry even with wait time risk of missing a settlement pair if a service bus outage occurs or the managed identity token refresh happens between them and takes a while.

All very low risk though.  What do you think?  Keep it or drop it?